### PR TITLE
Add iputils

### PR DESCRIPTION
**Pull request by andreistan26**
#20 Added iputils

### CODE REVIEW

**Review by StefanJum — CHANGES_REQUESTED** [review]
This looks good, please add a `README.md` file where you specify how to run the apps (both under Linux and under Unikraft, using `run.sh`).

Also please update the commit message and description to something more descriptive (see the [Unikraft contributing guidelines](https://unikraft.org/docs/contributing/submitting-changes/#commits)).

File: iputils/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+clockdiff 8.8.8.8
+ping 8.8.8.8
+tracepath 8.8.8.8
+```
+## Running on Unikraft
+All of the applications need networking so we need to use the option `-n` we can use the following command to run any of the binaries. This example is for ping:
+```
+sudo ./run.sh -n -k app-elfloader_qemu-x86_64 -r ./dynamic-apps/iputils ./bin/ping 8.8.8.8
+```
+
```
Comment (StefanJum) [inline iputils/README.md:16]:
```suggestion
```

File: iputils/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+Run the binaries using (for example) the following commands
+```
+arping 8.8.8.8
+clockdiff 8.8.8.8
+ping 8.8.8.8
+tracepath 8.8.8.8
+```
+## Running on Unikraft
+All of the applications need networking so we need to use the option `-n` we can use the following command to run any of the binaries. This example is for ping:
+```
```
Comment (StefanJum) [inline iputils/README.md:13]:
```suggestion
All of the applications need networking so we need to use the option `-n` we can use the following command to run any of the binaries. This example is for ping:

```console
```

File: iputils/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+## Running on Linux
+Run the binaries using (for example) the following commands
+```
+arping 8.8.8.8
+clockdiff 8.8.8.8
+ping 8.8.8.8
+tracepath 8.8.8.8
+```
+## Running on Unikraft
```
Comment (StefanJum) [inline iputils/README.md:11]: Add a new line before and after code snippets

File: iputils/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Running iputils
+
+## Running on Linux
+Run the binaries using (for example) the following commands
+```
```
Comment (StefanJum) [inline iputils/README.md:5]: Add a new line before and after code snippets

File: iputils/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Running iputils
+
+## Running on Linux
+Run the binaries using (for example) the following commands
```
Comment (StefanJum) [inline iputils/README.md:4]: Add a new line before and after headers

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thank you @andreistan26, the REAMDE file needs to be a bit more detailed, see [this file](https://github.com/unikraft/dynamic-apps/blob/877724017397574d435236c39defd1cefc6a7e68/lang/lua/README.md) for reference

File: iputils/README.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Running iputils
```
Comment (razvand) [inline iputils/README.md:1]: Add intro about the tools and the purpose of this directory. See the [`README.md` file in the `lang/lua/` directory](https://github.com/unikraft/dynamic-apps/blob/master/lang/lua/README.md).

Comment (razvand) [reply iputils/README.md:1]: Also add similar contents and wording.

File: iputils/README.md (hunk 4 above)
Comment (razvand) [inline iputils/README.md:5]: Add space before and after code snippets.

Comment (razvand) [reply iputils/README.md:5]: Mark `console` type for snippets, after triple backticks.

File: iputils/README.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+## Running on Linux
+Run the binaries using (for example) the following commands
+```
+arping 8.8.8.8
+clockdiff 8.8.8.8
+ping 8.8.8.8
+tracepath 8.8.8.8
+```
+## Running on Unikraft
+All of the applications need networking so we need to use the option `-n` we can use the following command to run any of the binaries. This example is for ping:
```
Comment (razvand) [inline iputils/README.md:12]:
```suggestion
All of the applications need networking so we need to use the option `-n`.
We can use the following command to run any of the binaries.
This example is for `ping`:
```

Comment (razvand) [reply iputils/README.md:12]: One sentence per line.

**Review by razvand — CHANGES_REQUESTED** [review]

File: iputils/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# Iputils Dynamic Build
```
Comment (razvand) [inline iputils/README.md:1]:
```suggestion
# iputils as Dynamic (PIE) Executables

```

Comment (razvand) [reply iputils/README.md:1]: Add new line after section title.

File: iputils/README.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+The **iputils** package is a set of small utilities for Linux networking.
+This directory includes the dynamic application build, that includes, the
+binary as well as the library dependencies for running inside of Unikraft.
+
+## Run Iputils Natively on Linux
+Run the binaries using (for example) the following commands
+
+``` console
```
Comment (razvand) [inline iputils/README.md:9]: Remove empty space between backtick and `console`.

File: iputils/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Iputils Dynamic Build
+The **iputils** package is a set of small utilities for Linux networking.
+This directory includes the dynamic application build, that includes, the
+binary as well as the library dependencies for running inside of Unikraft.
```
Comment (razvand) [inline iputils/README.md:4]: Add wording as in the [`README.md` file for `lang/lua`](https://github.com/unikraft/dynamic-apps/blob/master/lang/lua/README.md#lua-as-dynamic-pie-executable).

File: iputils/README.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+## Run Iputils Natively on Linux
```
Comment (razvand) [inline iputils/README.md:6]:
```suggestion
## Run iputils Natively on Linux

```

File: iputils/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+``` console
+arping 8.8.8.8
+clockdiff 8.8.8.8
+ping 8.8.8.8
+tracepath 8.8.8.8
+```
+
+## Run Iputils with Unikraft
+
+All of the applications need networking so we need to use the option `-n`.
```
Comment (razvand) [inline iputils/README.md:18]:
```suggestion
To run the binary ELF on Unikraft, run the `run.sh` script from the [`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader):
All of the applications need networking so we need to use the option `-n`.
```

File: iputils/README.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+Run the binaries using (for example) the following commands
+
+``` console
+arping 8.8.8.8
+clockdiff 8.8.8.8
+ping 8.8.8.8
+tracepath 8.8.8.8
+```
+
+## Run Iputils with Unikraft
```
Comment (razvand) [inline iputils/README.md:16]:
```suggestion
## Run iputils with Unikraft
```

File: iputils/README.md [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+All of the applications need networking so we need to use the option `-n`.
+We can use the following command to run any of the binaries.
+This example is for `ping`:
+
+``` console
```
Comment (razvand) [inline iputils/README.md:22]: Remove space between backticks and `console`.

File: iputils/README.md (hunk 10 above)
Comment (razvand) [reply iputils/README.md:4]: Do a listing of the executables and their purpose.

**Review by razvand — CHANGES_REQUESTED** [review]
@andreistan26, see my comments. Take a peek into the [`README.md` file for `lang/lua`](https://github.com/unikraft/dynamic-apps/blob/master/lang/lua/README.md).

**Comment by andreistan26** [comment]
Definitely looks better now, should be good to go for another review.

File: iputils/README.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+## Run iputils with Unikraft
+
+To run the binary ELF on Unikraft, run the `run.sh` script from the [`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader):
+
+All of the applications need networking so we have to use the option `-n`.
+We can use the following command to run any of the binaries.
+This example is for `ping`:
+
+```console 
+sudo ./run.sh -n -k app-elfloader_qemu-x86_64 -r ./dynamic-apps/iputils ./bin/ping 8.8.8.8
```
Comment (StefanJum) [inline iputils/README.md:35]:
```suggestion
sudo ./run.sh -n -k app-elfloader_qemu-x86_64 -r ../dynamic-apps/iputils ./bin/ping 8.8.8.8
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Only one comment @andreistan26, all good besides that

File: iputils/README.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+tracepath 8.8.8.8
+arping 8.8.8.8
+clockdiff 8.8.8.8
+```
+
+## Run iputils with Unikraft
+
+To run the binary ELF on Unikraft, run the `run.sh` script from the [`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader):
+
+All of the applications need networking so we have to use the option `-n`.
```
Comment (StefanJum) [inline iputils/README.md:30]:
```suggestion
All the applications need networking, so we have to use the option `-n`.
```

File: iputils/README.md [hunk 17]
```diff
@@ -0,0 +1,36 @@
+# iputils as Dynamic (PIE) Executables
+
+This directory stores the binary executable (ELF - Executable and Linkable Format), dynamic library dependencies and support files to run the applications provided by iputils. 
+The binary is PIE (Position-Independent Executable).
+All files are stored in their expected location in the Linux filesytem hierarchy.
```
Comment (StefanJum) [inline iputils/README.md:5]:
```suggestion
All files are stored in their expected location from the Linux filesytem hierarchy. 
```

**Review by StefanJum — APPROVED** [review]
See the few minor comments, I'm all good after that.

@razvand please also take a look over this.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

File: iputils/README.md [hunk 18]
```diff
@@ -0,0 +1,36 @@
+# iputils as Dynamic (PIE) Executables
+
+This directory stores the binary executable (ELF - Executable and Linkable Format), dynamic library dependencies and support files to run the applications provided by iputils.
```
Comment (razvand) [inline iputils/README.md:3]:
```suggestion
This directory stores the binary executable (ELF - Executable and Linkable Format), dynamic library dependencies and support files to run the applications provided by iputils.
```

File: iputils/README.md [hunk 19]
```diff
@@ -0,0 +1,36 @@
+# iputils as Dynamic (PIE) Executables
+
+This directory stores the binary executable (ELF - Executable and Linkable Format), dynamic library dependencies and support files to run the applications provided by iputils. 
+The binary is PIE (Position-Independent Executable).
+All files are stored in their expected location from the Linux filesytem hierarchy.
```
Comment (razvand) [inline iputils/README.md:5]:
```suggestion
All files are stored in their expected location from the Linux filesytem hierarchy.
```

File: iputils/README.md [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+## Run iputils with Unikraft
+
+To run the binary ELF on Unikraft, run the `run.sh` script from the [`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader):
+
+All the applications need networking, so we have to use the option `-n`.
+We can use the following command to run any of the binaries.
+This example is for `ping`:
+
+```console
```
Comment (razvand) [inline iputils/README.md:34]: Removing trailing white space.

**Review by razvand — CHANGES_REQUESTED** [review]
@andreistan26, see my minor comments. Other than that, everything works. Excellent work!

After you apply my requested changes, I will then approve this and it will be upstream.

File: iputils/README.md [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+This directory stores the binary executable (ELF - Executable and Linkable Format), dynamic library dependencies and support files to run the applications provided by iputils. 
+The binary is PIE (Position-Independent Executable).
+All files are stored in their expected location from the Linux filesytem hierarchy. 
+The aim is to run the executable in any Linux ELF-compatible environment: native Linux, chroot/jail, Docker, Unikraft.
+
+## Tools Included in iputils
+
+- [ping](https://man7.org/linux/man-pages/man8/ping.8.html) - send ICMP requests to network hosts
+- [tracepath](https://man7.org/linux/man-pages/man8/tracepath.8.html) - traces path to a network host
```
Comment (razvand) [inline iputils/README.md:11]:
```suggestion
- [tracepath](https://man7.org/linux/man-pages/man8/tracepath.8.html) - trace path to a network host
```

**Review by razvand — CHANGES_REQUESTED** [review]

**Review by RaduNichita — APPROVED** [review]
All good from my side. Thanks, @andreistan26 !

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

**Pull request closed without merging**